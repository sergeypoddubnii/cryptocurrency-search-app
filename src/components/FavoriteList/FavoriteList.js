import React, { useMemo } from 'react';
import { useSelector } from 'react-redux';
import favCryptoSelectors from '../../redux/favCryptocurrency/favCryptoSelectors';
import FavoriteListItem from '../FavoriteListItem/FavoriteListItem';
import styled from 'styled-components';

const Title = styled.h2`
  font-size: 22px;
  padding: 0;
  margin-bottom: 10px;
`;

const FavoriteList = () => {
  const favCryptocurrencies = useSelector(
    favCryptoSelectors.getFavCryptocurrencies,
  );

  const cryptocurrenciesList = useMemo(() => {
    return favCryptocurrencies.map(crypto => {
      return (
        <FavoriteListItem
          key={crypto.id}
          name={crypto.name}
          symbol={crypto.symbol}
          price={crypto.price}
          id={crypto.id}
        />
      );
    });
  }, [favCryptocurrencies]);

  console.log(favCryptocurrencies);
  return (
    <div>
      <Title>Your favorite list</Title>
      <ul>{cryptocurrenciesList}</ul>
    </div>
  );
};

export default FavoriteList;
