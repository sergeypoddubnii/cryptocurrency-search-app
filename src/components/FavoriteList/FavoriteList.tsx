import React, { useMemo } from 'react';
import { useSelector } from 'react-redux';
import favCryptoSelectors from '../../redux/favCryptocurrency/favCryptoSelectors';
import FavoriteListItem from '../FavoriteListItem/FavoriteListItem';
import withTitleList from '../../hoc/withTitleList';

interface cryptoCurrency {
  id: string;
  name: string;
  symbol: any;
  price: number;
}

const hocOptions = { title: 'Your favorite cryptoсurrencies:' };

const FavoriteList = () => {
  const favCryptocurrencies = useSelector(
    favCryptoSelectors.getFavCryptocurrencies,
  );

  const cryptocurrenciesList = useMemo(() => {
    return favCryptocurrencies.map((crypto: cryptoCurrency) => {
      return (
        <FavoriteListItem
          key={crypto.id}
          name={crypto.name}
          symbol={crypto.symbol}
          price={crypto.price}
          id={crypto.id}
        />
      );
    });
  }, [favCryptocurrencies]);

  return <ul>{cryptocurrenciesList}</ul>;
};

export default withTitleList(hocOptions)(FavoriteList);
