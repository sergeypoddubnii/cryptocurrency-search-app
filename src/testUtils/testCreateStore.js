import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';

const testCreateStore = (store = {}) => configureStore([thunk])({ ...store });

export default testCreateStore;
