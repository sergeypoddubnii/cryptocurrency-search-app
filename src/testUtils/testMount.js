import React from 'react';
import { Provider } from 'react-redux';
import { mount } from 'enzyme';

const testMount = (Component, store = {}) => mount(<Provider store={store}>{Component}</Provider>);

export default testMount;
