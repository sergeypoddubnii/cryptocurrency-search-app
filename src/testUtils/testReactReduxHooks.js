import { useSelector, useDispatch } from 'react-redux';

const testReactReduxHooks = {
	useSelector: (state) => useSelector(state),
	useDispatch: () => useDispatch(),
};

export default testReactReduxHooks;
