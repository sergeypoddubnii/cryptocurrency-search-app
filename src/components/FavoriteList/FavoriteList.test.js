// eslint-disable-next-line no-use-before-define
import React from 'react';
import {
	shallow,
} from 'enzyme';
import FavoriteList from './FavoriteList.tsx';
import testMount from '../../testUtils/testMount';
import testReactReduxHooks from '../../testUtils/testReactReduxHooks';
import testCreateStore from '../../testUtils/testCreateStore';

const fakeCryptoStore = {
	favCryptocurrencies: [
		{
			id: 1, name: 1, symbol: 1, price: 1,
		},
		{
			id: 2, name: 2, symbol: 2, price: 2,
		},
		{
			id: 3, name: 3, symbol: 3, price: 3,
		},
	],
};

describe('FavoriteList component', () => {
	const favoriteList = shallow(<FavoriteList />);
	// mocking store
	const store = testCreateStore(fakeCryptoStore);
	// mock useSelector on mock store
	jest
		.spyOn(testReactReduxHooks, 'useSelector')
		.mockImplementation(() => store.getState());

	it('FavoriteList render properly', () => {
		expect(favoriteList).toMatchSnapshot();
	});

	it('FavoriteList render should render list of FavoriteListItem', () => {
		const favoriteList = testMount(<FavoriteList />, store);
		expect(favoriteList.find('FavoriteListItem')).toHaveLength(3);
	});
});
