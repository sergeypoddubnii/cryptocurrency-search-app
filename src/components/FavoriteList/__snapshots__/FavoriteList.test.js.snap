// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FavoriteList component FavoriteList render properly 1`] = `
<styled.div>
  <styled.h2>
    Your favorite crypto—Åurrencies:
  </styled.h2>
  <FavoriteList />
</styled.div>
`;
